### PR TITLE
The Inventory Rewrite

### PR DESCRIPTION
**Pull request by Camotoy**
All inventories should now be implemented.

Geyser's inventory translation has been rewritten to use Bedrock 1.16's new "server authoritative" inventory system.

Crafting should now be far better.

[ a lot more ]

NOTE TO MERGER: PERHAPS CONSIDER NOT SQUASH-MERGING.

### CODE REVIEW

File: bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java
Comment (Camotoy) [inline bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java:86]: Note to self: this probably isn't effective with Floodgate because it checks the session's Xbox authentication data and not the Java username. 

The problem is that, the older the version, the more likely that the packet that sends UUID and name information hasn't been sent yet. Perhaps that's when we can trigger sending recipes?

**Review by toinouH — COMMENTED** [review]
I have listed everything we could translate on #development of the Discord.

File: connector/src/main/java/org/geysermc/connector/network/translators/java/JavaDeclareRecipesTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/JavaDeclareRecipesTranslator.java:168]: Move this to the first line of the for loop. That makes it more clear that the method ends.

File: connector/src/main/java/org/geysermc/connector/utils/InventoryUtils.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/utils/InventoryUtils.java:135]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaTradeListTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaTradeListTranslator.java:56]: Add a few new lines in here

File: connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaWindowPropertyTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaWindowPropertyTranslator.java:44]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaWindowItemsTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaWindowItemsTranslator.java:45]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaSetSlotTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaSetSlotTranslator.java:73]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/updater/HorseInventoryUpdater.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/inventory/updater/HorseInventoryUpdater.java:58]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/PlayerInventoryTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/PlayerInventoryTranslator.java:207]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/PlayerInventoryTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/PlayerInventoryTranslator.java:110]: Missing spaces

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/CraftingInventoryTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/CraftingInventoryTranslator.java:43]: Please use brackets

File: connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockInventoryTransactionTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockInventoryTransactionTranslator.java:93]: Please use brackets

File: bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java
Comment (Tim203) [inline bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java:178]: The following code is quite compact. I personally put readability above line count.

File: connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaSetSlotTranslator.java
Comment (Tim203) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaSetSlotTranslator.java:162]: Can be in one if-statement

File: connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaSetSlotTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/java/window/JavaSetSlotTranslator.java:179]: Not quite sure how to go about that, because the variable is used later.

File: connector/src/main/java/org/geysermc/connector/inventory/PlayerInventory.java
Comment (AJ-Ferguson) [reply connector/src/main/java/org/geysermc/connector/inventory/PlayerInventory.java:62]: I was going to address this a while ago but forgot. As you said this problem extends to other inventory types too. If we are setting an item and the index is out of bounds, nothing should happen. And if we are getting an item and the index is out of bounds, we should return GeyserItemStack.EMPTY.

**Review by rtm516 — CHANGES_REQUESTED** [review]

File: bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/manager/GeyserSpigotWorldManager.java
Comment (Camotoy) [reply bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/manager/GeyserSpigotWorldManager.java:137]: `session.getPlayerEntity()` is a final variable that is always initialized.

File: connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockBookEditTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockBookEditTranslator.java:56]: I certainly don't care - but it really depends on if variable names are reused.

File: connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockLecternUpdateTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockLecternUpdateTranslator.java:59]: It might just be OK to set that to ZERO but I'll take a look and see if Java really cares before doing that.

File: connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockLecternUpdateTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockLecternUpdateTranslator.java:92]: It's how the button field in ClientClickWindowButtonPacket is transmitted over the network.

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/InventoryTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/inventory/InventoryTranslator.java:112]: Yes.

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/InventoryTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/inventory/InventoryTranslator.java:392]: I would disagree here because variables are reused within each switch statement, so the brackets prevent conflicts.

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/Generic3X3InventoryTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/Generic3X3InventoryTranslator.java:45]: The second parameter is for any other block that can also use this inventory - this is not a typo but intentional.

File: bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java
Comment (RednedEpic) [inline bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java:161]: Could this be stored as a constant? Unless there is something I'm missing, it seems very expensive to rerun this code every time a player joins.

File: connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockLecternUpdateTranslator.java
Comment (rtm516) [reply connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockLecternUpdateTranslator.java:59]: If thats not changed before merge, can you expand the TODO.

File: connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/CartographyInventoryTranslator.java
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/translators/inventory/translators/CartographyInventoryTranslator.java:70]: Maybe use a switch statement here?

**Review by RednedEpic — CHANGES_REQUESTED** [review]

File: connector/src/main/java/org/geysermc/connector/network/translators/item/RecipeRegistry.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/item/RecipeRegistry.java:222]: I'll take ideas, but the two are ever-so-slightly different - one adds a shapeless recipe, and the other adds a shaped recipe.

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/BlockTranslator.java
Comment (Camotoy) [reply connector/src/main/java/org/geysermc/connector/network/translators/world/block/BlockTranslator.java:115]: Well this is not supposed to be here! But yeah, I can explain.

File: connector/src/main/java/org/geysermc/connector/inventory/GeyserEnchantOption.java
Comment (RednedEpic) [reply connector/src/main/java/org/geysermc/connector/inventory/GeyserEnchantOption.java:69]: Will revisit later.

**Review by RednedEpic — APPROVED** [review]

**Comment by RednedEpic** [comment]
Many of the remaining suggestions/nitpicks I'm going to mark as resolved for now as they can be visited at a future date & are not blocking the release of this. Will be addressed at a future date as they likely apply to other classes in the codebase as well.

**Pull request merged**
